### ISSUE TITLE
Incompatibilities with numpy 2.0

### ISSUE THREAD

**Issue by jameswilburlewis** [spedas/pyspedas#1045]
It looks like our dependencies are now all compatible with numpy 2.0.0, so that's what we've started getting in our github test environment.   It turns out that `tplot_restore` is using a deprecated method`newbyteorder()` that is now expired:

```
Traceback (most recent call last):
  File "/home/runner/work/pyspedas/pyspedas/pyspedas/projects/elfin/tests/test_epd_l1.py", line 42, in setUpClass
    tplot_restore(filename)
  File "/opt/hostedtoolcache/Python/3.12.7/x64/lib/python3.12/site-packages/pytplot/importers/tplot_restore.py", line 137, in tplot_restore
    temp_x_data = temp_x_data.byteswap().newbyteorder()
                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
AttributeError: `newbyteorder` was removed from the ndarray class in NumPy 2.0. Use `arr.view(arr.dtype.newbyteorder(order))` instead.


```
py.org/doc/stable/numpy_2_0_migration_guide.html#ndarray-and-scalar-methods

I've pinned the numpy dependency in pyspedas requirements.txt to < 2.0.0 until we can resolve this (hopefully quickly, because upper limit pins are evil...)

**Comment by jameswilburlewis** [comment]
ChatGPT suggests that the numpy 2.0.0+ way of writing these calls is:


```
# Change from big endian to little endian since pandas apparently hates big endian
# We might want to move this into the store_data procedure eventually

if temp_x_data.dtype.byteorder == '>':
    temp_x_data = temp_x_data.byteswap().view(temp_x_data.dtype.newbyteorder())
if temp_y_data.dtype.byteorder == '>':
    temp_y_data = temp_y_data.byteswap().view(temp_y_data.dtype.newbyteorder())
if temp_v1_data.dtype.byteorder == '>':
    temp_v1_data = temp_v1_data.byteswap().view(temp_v1_data.dtype.newbyteorder())
if temp_v2_data.dtype.byteorder == '>':
    temp_v2_data = temp_v2_data.byteswap().view(temp_v2_data.dtype.newbyteorder())
if temp_v3_data.dtype.byteorder == '>':
    temp_v3_data = temp_v3_data.byteswap().view(temp_v3_data.dtype.newbyteorder())
```

The `byteswap()` call changes the endianness, but doesn't update the metadata.  The `view(arr.dtype.newbyteorder())` call ensures that the dtype metadata in the output matches what is actually stored in the variable.

**Comment by jameswilburlewis** [comment]
With the newbyteorder() patch in place, we're still getting failures in the THEMIS spinmodel test suite when I let it use numpy >= 2.0.0 (using python 3.12).   Not sure yet if it's tplot_restore, or some other routine that has a problem....

**Comment by jameswilburlewis** [comment]
The problem we were having had to do with new type promotion rules in numpy 2.0, when performing operations on mixed numpy and Python scalar types.    The spinmodel segments have integer valued fields c1 and c2 (start and end spin numbers for each spin model segment).  It looks like they come out of the CDF as 16-bit unsigned integers.   At UTC date boundaries, we usually end up doing some arithmetic that mixes numpy uint16 values and Python integer literals.  

Prior to numpy 2, this caused a promotion to int64, which propagated to the rest of the spinmodel segments.  With numpy 2, the result remained uint16.  Later in the processing, we end up trying to add two np.uint16 values that overflowed, and the result wrapped around mod 65536 instead of being promoted to a larger integer type.  So the spin model was nonsense after that point.

I was able to address this by explicitly converting the segment c1 and c2 values to np.int32 each time we construct a new segment.  With that patch in place, the spinmodel tests pass under numpy 2+.

Hopefully this will be a rare situation.  We did at least get a warning about the overflow when using numpy 2, so by checking the test logs, we can find any other similar errors.
